PR-02-B: Flink REST Client.

rdziadosz opened the pull request:
See the project plan at [#439](https://github.com/delta-io/connectors/issues/439).
### What is the purpose of the change
Flink cluster is used to execute all test jobs. JobManager REST API is exposed, so the test framework can communicate with it. REST Client enables scheduling and managing Flink jobs during tests.

CustomRestClient was introduced so that Delta connector can properly commit new files (the checkpoint is triggered at the end of the job). In addition, it allows jar upload and job triggering using two separate methods.
### Verifying this change
The easiest way to verify is to start a local flink cluster and run example jobs on it.
The following code snippet runs an example job WordCount. Remember to set the path to the jar file as well as the port and host of jobmanager properly.
The following code was tested with a standalone flink cluster. If you are running flink on a docker the code below may require a few changes (access to input/output files).
```
package io.delta.flink.e2e;

import java.time.Duration;
import java.time.Instant;

import io.delta.flink.e2e.client.FlinkClient;
import io.delta.flink.e2e.client.FlinkClientFactory;
import io.delta.flink.e2e.client.parameters.JobParameters;
import io.delta.flink.e2e.client.parameters.JobParametersBuilder;
import org.apache.flink.api.common.JobID;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;

public class FlinkClientTest {
    // host on which the Flink API is available
    private static final String host = "localhost";
    // port on which the Flink API is available
    private static final int port = 8081;

    // path to WordCount example
    private static final String jarPath = "/Users/radoslaw.dziadosz/flink/flink-1.15.2/examples/batch/WordCount.jar";

    // Please note that the JobManager must have access to the following files:
    // path to WordCount input - it can be any text file
    private static final String wordCountInput= "/Users/radoslaw.dziadosz/flink/flink-1.15.2/LICENSE";
    // path to WordCount output - example job results will be written to this file
    private static final String wordCountOutput = "/tmp/flink_test_" + System.currentTimeMillis() + ".txt";

    private final JobParametersBuilder wordCountJob = JobParametersBuilder
            .builder()
            .withParallelism(1)
            .withEntryPointClassName("org.apache.flink.examples.java.wordcount.WordCount")
            .withArgument("input", wordCountInput) //path to any text file to be used as WorldCount input
            .withArgument("output", wordCountOutput);

    @Test
    @DisplayName("CustomRestClient should run an example job")
    public void flinkCustomClientTest() throws Exception {
        FlinkClient flinkClient = FlinkClientFactory.getCustomRestClient(host, port);
        String jarId = flinkClient.uploadJar(jarPath);

        JobParameters jobParameters = wordCountJob
                .withJarId(jarId)
                .build();

        JobID jobID = flinkClient.run(jobParameters);

        wait(flinkClient, jobID, Duration.ofSeconds(20));
        flinkClient.deleteJar(jarId);
    }

    private void wait(FlinkClient flinkClient, JobID jobID, Duration waitTime) throws Exception {
        Instant waitUntil = Instant.now().plus(waitTime);
        while (!flinkClient.isFinished(jobID) && Instant.now().isBefore(waitUntil)) {
            if (flinkClient.isFailed(jobID) || flinkClient.isCanceled(jobID)) {
                Assertions.fail(String.format("Job has failed or has been cancelled; status=%s.", flinkClient.getStatus(jobID)));
            }
            Thread.sleep(1_000L);
        }
        Assertions.assertTrue(flinkClient.isFinished(jobID), "Job has not finished in a timely manner.");
    }
}
```

Expected output:
```
14:54:22,164 INFO  i.d.f.e.c.CustomRestClient [] - Uploading jar: /Users/radoslaw.dziadosz/flink/flink-1.15.2/examples/batch/WordCount.jar
14:54:22,247 INFO  i.d.f.e.c.CustomRestClient [] - Jar uploaded; id=ef769a63-909d-49b4-ae6a-4cb5fe71432a_WordCount.jar
14:54:22,251 INFO  i.d.f.e.c.CustomRestClient [] - Submitting flink job; parameters: JobParameters{jarId='ef769a63-909d-49b4-ae6a-4cb5fe71432a_WordCount.jar', entryPointClassName='org.apache.flink.examples.java.wordcount.WordCount', parallelism=1, arguments={output=/tmp/flink_test_1669643661561.txt, input=/Users/radoslaw.dziadosz/flink/flink-1.15.2/LICENSE}}
14:54:22,341 INFO  i.d.f.e.c.CustomRestClient [] - Job submitted; jobId=b3ccd68c4708b2fcea0970d24d4dca7b.
14:54:25,409 INFO  i.d.f.e.c.CustomRestClient [] - Deleting jar ef769a63-909d-49b4-ae6a-4cb5fe71432a_WordCount.jar.
14:54:25,414 INFO  i.d.f.e.c.CustomRestClient [] - Jar ef769a63-909d-49b4-ae6a-4cb5fe71432a_WordCount.jar deleted.
```

scottsand-db commented:
Can you please update the PR description @rdziadosz ?

kristoffSC commented on flink/end-to-end-tests/src/test/java/io/delta/flink/e2e/client/CustomRestClient.java:
```diff
@@ -0,0 +1,240 @@
+
+    private final String host;
+    private final int port;
+    private final OkHttpClient httpClient;
```
Hi,
Any reason to not use Java 11 HttpClient?
We could dropped "com.squareup.okhttp3" dependency then.

rdziadosz replied:
Yes, okhttp has a better URL model. If the additional dependency is a problem, we can change it to Java HttpClient.

scottsand-db replied:
I'm fine with OkHttpClient.

kristoffSC commented on flink/end-to-end-tests/src/test/java/io/delta/flink/e2e/client/CustomRestClient.java:
```diff
@@ -0,0 +1,240 @@
+ */
+class CustomRestClient implements FlinkClient {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CustomRestClient.class);
```
nit:
Typical convention is to use "LOG" rather then "LOGGER".

rdziadosz replied:
Changed

kristoffSC commented on flink/end-to-end-tests/src/test/java/io/delta/flink/e2e/client/CustomRestClient.java:
```diff
@@ -0,0 +1,240 @@
+            return JobID.fromHexString(jobId);
+        } catch (Exception e) {
+            LOGGER.error("Failed to submit job.", e);
+            throw e;
```
https://www.theserverside.com/tip/Troubleshooting-Java-Code-Log-or-Re-Throw-but-Dont-Do-Both

rdziadosz replied:
I removed the unnecessary log

kristoffSC replied:
Ok but do you need this try/catch at all now?
the `run` is already marked as `throws Exception` and since you dont do anything in catch block, then why have it at all?

rdziadosz replied:
Of course we don't need it. Removed

kristoffSC commented on flink/end-to-end-tests/src/test/java/io/delta/flink/e2e/client/CustomRestClient.java:
```diff
@@ -0,0 +1,240 @@
+            if (!response.isSuccessful()) {
+                LOGGER.error("Request failed; code={}; body={}", response.code(),
+                        body == null ? null : body.string());
+                throw new RuntimeException("Request failed: " + response);
```
https://www.theserverside.com/tip/Troubleshooting-Java-Code-Log-or-Re-Throw-but-Dont-Do-Both

rdziadosz replied:
I removed the unnecessary log

kristoffSC commented on flink/end-to-end-tests/src/test/java/io/delta/flink/e2e/client/FlinkRestClient.java:
```diff
@@ -0,0 +1,124 @@
+ */
+class FlinkRestClient implements FlinkClient {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FlinkRestClient.class);
```
I would suggest "LOG" instead "LOGGER"

rdziadosz replied:
Changed

kristoffSC commented on flink/end-to-end-tests/src/test/java/io/delta/flink/e2e/client/parameters/JobParameters.java:
```diff
@@ -0,0 +1,72 @@
+
+import java.util.Map;
+
+public class JobParameters {
```
constructor is package protected but class is public, whys that?

kristoffSC commented:
Hi,
I've did a first review round and left few comments about variable (LOGGER) names and exception handling.

I think that what could be added herte is couple unit tests that will use WireMock [1] that we use for testing the REST client.

You can check how I use Wiremock (usage and setup) for connector tests it in my other project:
https://github.com/getindata/flink-http-connector/blob/main/src/test/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSourceITCaseTest.java

https://github.com/getindata/flink-http-connector/blob/main/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkConnectionTest.java


[1] https://wiremock.org/

kristoffSC commented:
Hi @rdziadosz 
I ran your snipped and it seems to be working, at least for custom Rest Client, Thanks :)

One question,
whats up with default Flink Rest client? Will wee use it? I guess not since there is a Custom client build. 
Could you share why we need a custom Rest Client?

Also when I ran the second test, the one that is using Default Rest Client i got this exception in Jon Manager logs (job failed)
`Caused by: org.apache.flink.api.common.InvalidProgramException: The job graph is cyclic.`

The same job (WordCount.jar) was working fine in the first test (see screen below). 
I'm runing this on Flink 1.15.2.

kristoffSC commented:
@rdziadosz 
I also used run-end-to-end-tests.sh script to build and run tests on local flink cluster. 
Everything worked well, this is great :) 

Like we discussed offline, I think it would be a good thong to add a flag to run-end-to-end-tests.sh that will allow to use local Flink cluster and turn off all terraform stuff, like I had to to manually. 

In short, LGTM -> works from IDE running individual tests + using sh. script. :) 
I have not tested this with terraform setup though. 

Please address my previous comments + I would like to kindly ask you to add javadoc for all classes.

rdziadosz commented on flink/end-to-end-tests/src/test/java/io/delta/flink/e2e/client/CustomRestClient.java:
```diff
@@ -0,0 +1,240 @@
+import java.util.stream.Collectors;
+
+import io.delta.flink.e2e.client.parameters.JobParameters;
+import okhttp3.*;
```
I got rid of the '*' from the imports

rdziadosz commented:
Hi @kristoffSC,
Thank you for the review.

I pushed new changes to this pull request. I added javadoc and created few tests using WireMock - thanks for the examples.

I also removed the client that uses RestClusterClient. We don't need it as we have CustomRestClient. CustomRestClient was introduced because RestClusterClient is intended for internal library usage. When using this client, we experienced problems with committing new files to Delta tables. It also forces you to upload a jar file every time you run the Flink job, which is slow.

You experienced a problem with `The job graph is cyclic,` due to Flink version incompatibility. On this branch, we still have version 1.14. I created a new pull request to our feature branch with new changes from master branch and with Flink version upgrade to 1.15: https://github.com/delta-io/connectors/pull/480

vkorukanti commented:
This repo has been deprecated and the code is moved under `connectors` module in https://github.com/delta-io/delta repository. [Here](https://github.com/delta-io/connectors/issues/556#issuecomment-1631207507) are the migration steps to recreate this PR in the new repository location.
